### PR TITLE
Add macros for ship designs/hulls/parts and monsters

### PR DESCRIPTION
**Pull request by Ouaz**
... in order to be produced on a given location.

- For a convenient reading directly in the stringtables, keys are built by mixing the corresponding building keys:
 
Ex: BLD_SHIPYARD_TYPE1_TYPE2_TYPE3_REQUIRED

```
BLD_SHIPYARD_BASE_ORBITAL_DRYDOCK_CON_GEOINT_REQUIRED
= Requires a Basic Shipyard, an Orbital Drydock, and a Geo-Integration Facility
```

### CODE REVIEW

**Comment by geoffthemedio** [comment]
any particular feedback requested?

**Comment by MatGB** [comment]
Looks fine to me and is a good idea, standardising text as much as possible is something we should try to do, not least for error reduction and translator workload reduction.

**Comment by Ouaz** [comment]
I ask here: I wanted to do the same thing for ship hull descriptions (in another PR), where there is several times the same recurring sentences.

For example, for Asteroid Hulls:

In SH_SMALL_ASTEROID_DESC:

`Requires a [[buildingtype BLD_SHIPYARD_BASE]] and an asteroid belt with an [[buildingtype BLD_SHIPYARD_AST]] present in the system where it is produced.`

which is now an equivalent of:

```
BLD_SHIPYARD_BASE_AST_REQUIRED
Can only be built at a colony with a [[buildingtype BLD_SHIPYARD_BASE]] in a system 
with an [[buildingtype BLD_SHIPYARD_AST]].
```

Should I create new macro keys to be used specifically in Ship Hull descriptions, or re-use the Ship Designs macro keys (with some modification if needed)?

For example:

```
BLD_SHIPYARD_BASE_AST_REQUIRED
Can only be built at a colony with a [[buildingtype BLD_SHIPYARD_BASE]] in a system
with an asteroid belt and a [[buildingtype BLD_SHIPYARD_AST]].
```

**Comment by MatGB** [comment]
If it's for the same hull then I think yes, it'll be far easier to change/update in the future if all the premade designs use the same text as the hulls.

**Comment by Ouaz** [comment]
> If it's for the same hull then I think yes, it'll be far easier to change/update in the future if all the premade designs use the same text as the hulls.

Ok, done.
- In the macro keys, I changed "_at a colony_" by a more generic "_at a location_", as some buildings can only be built on an asteroid belt, and others may be built on an outpost.
- I fixed the Fractal ship design description, as it can only be built at a system with Blue or Black Hole star types (not White).

I have a doubt about the Asteroid Reformation Processor. In the ship design description, it says:
`An [[buildingtype BLD_SHIPYARD_AST_REF]] is also required in any system owned by the empire.`

but in the hull description, it says:
`and an [[buildingtype BLD_SHIPYARD_AST_REF]] present in the system where it is produced.`

Which is right? (I kept the first in the macro key)

Thanks.

**Comment by dbenage-cx** [comment]
For `BLD_SHIPYARD_AST_REF`:
Some designs like `SD_AST_3` have parts (`AR_ROCK_PLATE`) that require this building in the system, owned by the empire or by an ally.
For `SD_CRYSTAL`, the hull requires the building to be owned by the empire (ally does not count).
This design also has `AR_CRYSTAL_PLATE` with the same requirement of empire owned only.

**Comment by Ouaz** [comment]
OK, thanks.

If I check the designs or hulls requiring a `BLD_SHIPYARD_AST_REF` one by one, and adapt the macro keys accordingly, it gives:

```
BLD_SHIPYARD_AST_REF_REQUIRED_SAME_SYSTEM
An [[buildingtype BLD_SHIPYARD_AST_REF]] is also required in the same system
owned by the empire or by an ally.

BLD_SHIPYARD_AST_REF_REQUIRED_SAME_SYSTEM_EMPIRE_ONLY
An [[buildingtype BLD_SHIPYARD_AST_REF]] is also required in the same system
owned by the empire.

BLD_SHIPYARD_AST_REF_REQUIRED_OWNED_SYSTEM
An [[buildingtype BLD_SHIPYARD_AST_REF]] is also required in any system
owned by the empire or by an ally.

BLD_SHIPYARD_AST_REF_REQUIRED_OWNED_SYSTEM_EMPIRE_ONLY
An [[buildingtype BLD_SHIPYARD_AST_REF]] is also required in any system
owned by the empire.
```

Is this OK?

**Comment by geoffthemedio** [comment]
I'd rephrase to "An AST REF owned by the empire [or an ally] is required [in the same system]" to make it clear that it's the AST REF that needs to be owned, not the system.

**Comment by Ouaz** [comment]
Here's the list of all items (ship designs or hulls - I will make ship parts later), requiring an Asteroid Reformation Processor:

```
SD_AST_3  (Rock Destroyer P2)
SD_HEAVY_AST_2 (Rock Battleship P)
SD_CRYSTAL (Crystal Destroyer P)
SH_AGREGATE_ASTEROID (Aggregate Asteroid Hull)
SH_MINIASTEROID_SWARM (Mini-Asteroid Swarm)
SH_SCATTERED_ASTEROID (Scattered Asteroid Hull)
SH_CRYSTALLIZED_ASTEROID (Crystallized Asteroid Hull)
```

Something I still don't understand (because it doesn't match with what dbenage-cx says in [his comment](https://github.com/freeorion/freeorion/pull/1095#issuecomment-260249433)):

The Rock Armor Plating (`AR_ROCK_PLATE`) description says (as the Crystal Armor Plating):

`This part can only be built if there is an Asteroid Reformation Processor at an asteroid belt somewhere in the empire or an ally's empire.`

Does this mean, for example, that a `SD_AST_3` (Rock Destroyer P2, which requires a `AR_ROCK_PLATE`) can be built like this:
- Empire's planet in a system: Basic Shipyard
- Empire's asteroid belt in the same system: Asteroid Processor
- Empire or Ally's asteroid belt in another system: Asteroid Reformation Processor

Or everything has to be in the same system? (Whatever is the owner, empire or ally when relevant). Same questions for all items in the list.

Because, if I check each item description (and the focs), I now have only two possible macros:

For `SD_AST_3`  (Rock Destroyer P2), `SD_HEAVY_AST_2` (Rock Battleship P), `SD_CRYSTAL` (Crystal Destroyer P)

```
BLD_SHIPYARD_AST_REF_REQUIRED_ANY_SYSTEM
An [[buildingtype BLD_SHIPYARD_AST_REF]] owned by the empire or by an ally
is also required in any system.

## Or in the same system?
```

For `SH_AGREGATE_ASTEROID` (Aggregate Asteroid Hull), `SH_MINIASTEROID_SWARM` (Mini-Asteroid Swarm), `SH_SCATTERED_ASTEROID` (Scattered Asteroid Hull), `SH_CRYSTALLIZED_ASTEROID` (Crystallized Asteroid Hull):

```
BLD_SHIPYARD_AST_REF_REQUIRED_SAME_SYSTEM_EMPIRE_ONLY
An [[buildingtype BLD_SHIPYARD_AST_REF]] owned by the empire
is also required in the same system.

## Or in any system?
```

**Comment by dbenage-cx** [comment]
Not sure what I was looking at earlier, both of these look correct with the exception of `SD_CRYSTAL` (it uses `SH_CRYSTALLIZED_ASTEROID`).
screenshot_20161116_163538

**Comment by geoffthemedio** [comment]
For AR_ROCK_PLATE the location condition is
https://github.com/freeorion/freeorion/blob/master/default/scripting/ship_parts/Armor.focs.txt#L53

```
    And [
        Number low = 1 high = 999 condition = And [
            Building name = "BLD_SHIPYARD_AST_REF"
            Or [
                OwnedBy empire = Source.Owner
                OwnedBy affiliation = AllyOf empire = Source.Owner
            ]
        ]
        OwnedBy empire = Source.Owner
    ]
```

That means the location must be owned by the empire producing the armour part ship design, and there needs to be at least one asteroid reformation processor owned by the same empire or owned by an ally of that empire.

**Comment by Ouaz** [comment]
OK, thanks, I'm on it.

As there's now a huge pending PR by dbenage-cx (#1112), I will wait it is merged before I update mine (because otherwise it will require a lot of merging conflict resolutions ^^)

**Comment by dbenage-cx** [comment]
@Ouaz sorry for the mess, I was not thinking about conflicts with stringtable.  If you prefer, I can resolve the conflicts.

**Comment by Ouaz** [comment]
> so I guess you guys can proceed here...

I have still to insert some macros in the Predefined Ship Designs section. I waited dbenage-cx PR to be merged in order to have less merging conflicts. ^^

> sorry for the mess, I was not thinking about conflicts with stringtable. If you prefer, I can resolve the conflicts.

No problemo, your PR was way more important. ^^ If it's really a mess and I can't handle the conflicts, I will query your help. :p

**Comment by Ouaz** [comment]
Rebased and updated (Ship Designs and Monsters last macros added).

Works fine in latest build (2016-11-22.fdc2eea)
